Fix behavior of PM list with deactivated users

Signior-X opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes #13766
Follow-up PR #13795 
Rebase the PR and tested properly.
1. Allowing deactivated users pill in the compose box. thus prevents pm-list from collapsing.
2. Showing the deactivated user pill in red.
3. Showing compose error with deactivated user detail.

**Testing plan:** <!-- How have you tested? -->
Tested properly with node tests. Also tested manually as mentioned in the issue.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
When you try to send a message to a deactivated user.
![image](https://user-images.githubusercontent.com/56730716/123335937-cf845680-d562-11eb-9720-38def6db28d2.png)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

Signior-X commented:
![image](https://user-images.githubusercontent.com/56730716/123335910-c4c9c180-d562-11eb-9684-fb5f342ad73f.png)

timabbott commented on static/js/people.js:
```diff
@@ -730,7 +730,7 @@ export function is_valid_email_for_compose(email) {
         return false;
     }
-    return active_user_dict.has(person.user_id);
+    return true;
```
We should add a comment here noting that we need to allow deactivated users in compose so that one can attempt to reply to threads that contained them.

Signior-X replied:
Added the comment.

timabbott commented on static/js/input_pill.js:
```diff
@@ -96,25 +98,57 @@ export function create(opts) {
+                "title",
+                $t({defaultMessage: "Can't send messages to the deactivated user"}),
+            );
+            return $pill;
```
This function seems to mostly duplicate the original; could we just have `is_active` be a parameter to the "render a user pill" function?

Signior-X replied:
I have created an option `deactivated_user_pill` that handles this properly. Yes I was also feeling that the code was almost repeated.

timabbott commented on static/js/compose.js:
```diff
@@ -784,6 +785,26 @@ function validate_private_message() {
+        );
+
+        return false;
+    }
```
I think it might read more cleanly to just do this `return` inside the above `for` loop, since we don't display more than one.

Signior-X replied:
Okay sure. Didn't noticed this much. updating this.

Signior-X commented:
@timabbott I have updated the PR with the changes as suggested. Check this now.

timabbott commented on static/js/input_pill.js:
```diff
@@ -96,23 +97,30 @@ export function create(opts) {
 
             if (has_image) {
-                opts.img_src = item.img_src;
+                opts.img_src = payload.item.img_src;
```
Why does this code switch from referencing `item.foo` to `payload.item.foo`?  If we need to do that, it might be worth doing in its own commit for readability.

Signior-X replied:
This was actually done by the old PR. While rebasing didn't noticed this. I will undo this.

timabbott commented on static/js/input_pill.js:
```diff
@@ -96,23 +97,30 @@ export function create(opts) {
 
+            if (item.type === "user" && !people.is_person_active(item.user_id)) {
+                opts.deactivated_user_pill = true;
+                opts.display_value = payload.item.display_value + " (deactivated)";
```
I think this belongs in `user_pill.js` somewhere, not in this file, right?

Signior-X replied:
I need to understand more of a bit why that PR added this check here. I just updated the check to be more better. I will see and let you know.

Signior-X replied:
Yes, so I researched more about this. I understood the whole code how `input_pills.js` is just for handling pills and we pass different functions in it. I debugged and found the right place to add it. Add this code in `user_pill.js`

Signior-X replied:
Now updating the test is left for commit 2.

Signior-X commented:
@timabbott I have updated the PR as suggested by you. Moved the logic in `user_pills.js`. Also, updated the tests in the 2nd commit for feature.
This is now ready for review.

timabbott commented on static/templates/input_pill.hbs:
```diff
@@ -1,4 +1,4 @@
-<div class='pill' data-id='{{ id }}' tabindex=0>
+<div class='pill {{#if not_active}} deactivated-user-pill {{/if}}' {{#if not_active}} title="{{t 'Can\'t send messages to the deactivated user' }}" {{/if}} data-id='{{ id }}' tabindex=0>
```
This is a general template for pills everywhere in the app; I think we should make at least the `title` a parameter passed in from the caller.  And probably we should consider having a `deactivated-pill` class rather than `deactivated-user-pill`, since the styling seems like it'd be appropriate for a deactivated stream (or something) as well in the future.

Finally, we should call the variable `deactivated` rather than `not_active`, to avoid adding new terminology.

@m-e-l-u-h-a-n may have some ideas for how to do this nicely.

Signior-X replied:
I updated the name. However, I have removed `title` for now. As I think, certain pills having title and certain not will not be a good idea. So, we should pass the title to every pill. Will add it in some other commit.

m-e-l-u-h-a-n replied:
I think we should show title only for deactivated pills, as users may be curious to know why their color is different. From changes in the most recent push, it appears the title is same as the display value for normal pills so that does not add much information here 🤔.

m-e-l-u-h-a-n replied:
Is this only for recipient box pills? Because user pills can also be from
https://github.com/zulip/zulip/blob/87fc2bbe50da15ebcbecfb710e1ef928224bcaa0/static/js/user_pill.js#L60
in `user_pill.js`, which is not changed. 
Most of the pill widgets (Adding members to streams, user group setting, custom profile type with user) use this for creating pills, so if we want this behavior for pills of deactivated users to be the same everywhere we should change this as well to show "(deactivated)" in display value. Maybe we should wait for @timabbott's suggestion here.

Signior-X replied:
I have already updated the way. see 4th commit (latest commit) for what change I did. I searched for where these pills and used and added condition based on that.

Signior-X commented:
@timabbott Updated and commented above.

Signior-X commented:
@timabbott I have update the PR and added the commit where title attribute is added to `input_pill.hbs`.

timabbott commented on static/js/compose.js:
```diff
@@ -758,6 +759,19 @@ function validate_private_message() {
+        if (!people.is_person_active(user_id)) {
+            context = {email: people.get_by_user_id(user_id).email};
+            compose_error.show(
+                $t_html({defaultMessage: "The user {email} has been deactivated"}, context),
```
I changed this string, both for consistency and to add a `.` at the end of the error message.  Can you do a follow-up pass to make that `.` appear at the end of all of these?

timabbott commented:
I ended up dropping the last/title commit when merging, after moving the message "You can't send to deactivated users" to be what happens when you try to send a message; I think with that, the situation is pretty clear (as the red highlighting suggests having that user as a recipient is an error).

Merged, thanks @Signior-X and @m-e-l-u-h-a-n!  I like the way the deactivated tag is implemented here a lot.
